### DISCUSSION TITLE
Does it support RTL languages ​​like Arabic?

### DISCUSSION THREAD

**Discussion by fahadahmadshemul** [artkonekt/pdf-invoice#75] [Q&A]
It gives an error when I am trying to add ar.inc file in the inc/language folder

Error: iconv(): Detected an illegal character in input string

**Comment by fulopattila122** [comment]
I'm not aware that the package is being explicitly tested for RTL languages.
We neither have the knowledge of RTL languages nor the bandwidth to implement support, but pull requests are welcome and are regularly merged.

---

## Tips to Try

The `iconv()` error you're seeing, specifically "Detected an illegal character in input string," suggests that the input string being passed to `iconv()` contains characters that are either not valid for the specified encoding or cannot be converted to the target encoding. Since you're dealing with RTL languages like Arabic, this could be related to encoding mismatches, as Arabic uses specific encodings like UTF-8 or ISO-8859-6.

Here are some tips to troubleshoot and fix the issue:

### 1.Check the Encoding of the Input File (`ar.inc`)

Ensure that the `ar.inc` file is saved with the correct character encoding. Arabic text should typically be encoded in **UTF-8**, which supports a wide range of characters, including Arabic. If it's saved in another encoding (e.g., ISO-8859-1 or Windows-1252), `iconv()` might struggle with the conversion.

- Open the file and ensure it's saved as **UTF-8**.
- You can use tools like `file` (on Unix-based systems) to check the encoding:
   ```bash
   file -i inc/language/ar.inc
   ```

### 2. Specify the Correct Source and Target Encoding in `iconv()`

If the text you're converting is already in UTF-8, make sure that the source encoding is correctly specified when using `iconv()`. For instance, if the string is in UTF-8 but the function expects ISO-8859-1, it will raise an error.

Example:

```php
$converted = iconv('UTF-8', 'ISO-8859-1//TRANSLIT', $input);
```

Make sure the source encoding matches the encoding of the input string.

### 3. Use Error Suppression or Fallback Handling

If you’re uncertain about the encoding or dealing with user-generated content where encoding may vary, you can use `//IGNORE` in the `iconv()` call to silently discard characters that cannot be converted:

```php
$converted = iconv('UTF-8', 'ISO-8859-1//IGNORE', $input);
```

However, using `//IGNORE` may result in some characters being lost. To avoid data loss, you might want to validate or clean the input data before conversion.

### 4.Double-check Input Data for Invalid Characters

If you're passing a string that includes characters outside the expected range (e.g., control characters or other non-printable characters), `iconv()` may throw this error. You can clean up or validate the input string before passing it to `iconv()`.

Example:
```php
$input = preg_replace('/[^\x{0600}-\x{06FF}]/u', '', $input); // Remove non-Arabic characters
```

### 5. Make Sure the PDF Library Supports RTL Languages

Ensure that the underlying PDF library you're using for rendering invoices supports RTL text rendering. The **FPDF** library may need additional extensions.

### 6. Test with a Simple Arabic String

Test whether the issue is with the content of the `ar.inc` file or the library's handling of RTL languages by trying to generate a PDF with a very simple Arabic string, like "مرحبا" (Hello in Arabic), to see if the issue persists. If it works, the problem might be specific to the content or encoding in `ar.inc`.

### Summary of Steps

- Verify the encoding of the `ar.inc` file.
- Ensure `iconv()` is using the correct source and target encodings.
- Use error suppression or clean the input if necessary.
- Test with simple Arabic text to isolate the issue.
